src/main/java/org/trigon/hardware/misc/leds/AddressableLEDStrip.java

```diff
@@ -0,0 +1,182 @@
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.util.Color;
+
+import java.util.function.Supplier;
+
+public class AddressableLEDStrip extends LEDStrip {
+    private static AddressableLED LED;
+    private static AddressableLEDBuffer LED_BUFFER;
+    private int lastBreatheLED;
+    private double lastLEDMovementTime = 0;
+    private double rainbowFirstPixelHue = 0;
+    private boolean areLEDsOnForBlinking = false;
+    private boolean alternateColor = true;
+    private int amountOfColorFlowLEDs = 0;
+
+    public static void setLED(AddressableLED led) {
+        LED = led;
+    }
+
+    public static void setLEDBuffer(AddressableLEDBuffer ledBuffer) {
+        LED_BUFFER = ledBuffer;
+    }
+
+    public AddressableLEDStrip(boolean inverted, int numberOfLEDs, int indexOffset) {
+        super(inverted, numberOfLEDs, indexOffset);
+        resetLEDSettings();
+    }
+
+    @Override
+    public void periodic() {
+        LED.setData(LED_BUFFER);
+    }
+
+    @Override
+    void clearLEDColors() {
+        staticColor(Color.kBlack);
+    }
+
+    @Override
+    void blink(Color firstColor, Color secondColor, double blinkingIntervalSeconds) {
+        double currentTime = Timer.getFPGATimestamp();
+        if (currentTime - lastLEDMovementTime > blinkingIntervalSeconds) {
+            lastLEDMovementTime = currentTime;
+            areLEDsOnForBlinking = !areLEDsOnForBlinking;
+        }
+        if (areLEDsOnForBlinking)
+            staticColor(firstColor);
+        else
+            staticColor(secondColor);
+    }
+
+    @Override
+    void staticColor(Color color) {
+        setLEDColors(color, 0, numberOfLEDs - 1);
+    }
+
+    @Override
+    void breathe(Color color, int breathingLEDs, double cycleTimeSeconds, boolean shouldLoop, boolean inverted, LarsonAnimation.BounceMode bounceMode) {
+        clearLEDColors();
+        inverted = this.inverted != inverted;
+        double moveLEDTimeSeconds = cycleTimeSeconds / numberOfLEDs;
+        double currentTime = Timer.getFPGATimestamp();
+        if (currentTime - lastLEDMovementTime > moveLEDTimeSeconds) {
+            lastLEDMovementTime = currentTime;
+            if (inverted)
+                lastBreatheLED--;
+            else
+                lastBreatheLED++;
+        }
+        if (inverted ? (lastBreatheLED < indexOffset) : (lastBreatheLED >= numberOfLEDs + indexOffset)) {
+            if (!shouldLoop) {
+                getDefaultCommand().schedule();
+                return;
+            }
+            lastBreatheLED = inverted ? indexOffset + numberOfLEDs : indexOffset;
+        }
+        for (int i = 0; i < breathingLEDs; i++) {
+            if (lastBreatheLED - i >= indexOffset && lastBreatheLED - i < indexOffset + numberOfLEDs)
+                LED_BUFFER.setLED(lastBreatheLED - i, color);
+            else if (lastBreatheLED - i < indexOffset + numberOfLEDs)
+                LED_BUFFER.setLED(lastBreatheLED - i + numberOfLEDs, color);
+        }
```

Comment:
This needs to be cleaned. I wrote it and I still can't and don't want to be able to undertand it